File: sorting/quicksort/javascript/quickSort.js

function Sort(input)
{
  QuickSort(input,0,input.length-1);
  return input;
}

function QuickSort(input,start,end)
{
  if(start<end)
  {
    let pi = Partition(input,start,end);
    QuickSort(input,start,pi-1);
    QuickSort(input,pi+1,end);
  }
}

function Partition(input,start,end)
{
  let pivot = input[start];
  let lesserIndex = start+1;
  let greaterIndex = end;
  while(lesserIndex<greaterIndex)
  {
    while(lesserIndex<end && pivot>=input[lesserIndex])
    {
      lesserIndex++;
    }
    while(greaterIndex>start && pivot<input[greaterIndex])
    {
      greaterIndex--;
    }
    if(lesserIndex<greaterIndex)
    {
      let temp = input[lesserIndex];
      input[lesserIndex] = input[greaterIndex];
      input[greaterIndex] = temp;
    }
  }
  if(pivot>input[greaterIndex])
  {
    let temp = input[greaterIndex];
    input[greaterIndex] = pivot;
    input[start] = temp;
  }
}

module.exports={Sort};

File: sorting/quicksort/python/quickSort.py

def Sort(input: list) -> list:
    QuickSort(input,0,len(input)-1)
    return input

def QuickSort(input: list, start: int, end: int) -> list:
    if start<end:
        pi = partition(input,start,end)
        QuickSort(input,start,pi-1)
        QuickSort(input,pi+1,end)

def partition(input : list, start: int, end: int) -> int :
    pivot = input[start]
    i = start+1
    j = end
    while i<j:
        while pivot>=input[i] and i<end:
            i+=1
        while pivot<input[j] and j>start:
            j-=1

        if i<j:
            temp = input[i]
            input[i] = input[j]
            input[j] = temp

    #swap pivot with current end
    if pivot>input[j]:
        temp = input[j]
        input[j] = pivot
        input[start] = temp
    
    return j
